src/plugins/usage_collection/server/usage_counters/search/search.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type {
+  ISavedObjectsRepository,
+  SavedObjectsFindOptions,
+} from '@kbn/core-saved-objects-api-server';
+import {
+  serializeCounterKey,
+  type UsageCountersSavedObjectAttributes,
+  USAGE_COUNTERS_SAVED_OBJECT_TYPE,
+} from '../saved_objects';
+import type {
+  UsageCounterSnapshot,
+  UsageCountersSearchOptions,
+  UsageCountersSearchParams,
+  UsageCountersSearchResult,
+} from '../types';
+import { usageCountersSearchParamsToKueryFilter } from './search_utils';
+
+export async function searchUsageCounters(
+  repository: ISavedObjectsRepository,
+  params: UsageCountersSearchParams,
+  options: UsageCountersSearchOptions = {}
+): Promise<UsageCountersSearchResult> {
+  const { namespace: filterNamespace } = params;
+  const { perPage = 100, page = 1 } = options;
+
+  const filter = usageCountersSearchParamsToKueryFilter(params);
+
+  const findParams: SavedObjectsFindOptions = {
+    ...(filterNamespace && { namespaces: [filterNamespace] }),
+    type: USAGE_COUNTERS_SAVED_OBJECT_TYPE,
+    sortField: 'updated_at',
+    sortOrder: 'desc',
+    filter,
+    searchFields: [USAGE_COUNTERS_SAVED_OBJECT_TYPE],
+    perPage,
+    page,
+  };
+  const res = await repository.find<UsageCountersSavedObjectAttributes>(findParams);
+
+  const countersMap = new Map<string, UsageCounterSnapshot>();
+  res.saved_objects.forEach(({ attributes, updated_at: updatedAt, namespaces }) => {
```

Comment:
IMO, aggregations might better achieve what we're after here.

I imagine an API like:

```
usageCounters.search({
  filters: {
    domainId,
    counterName,
    counterType,
    source,
    timestamp: { [lt|lte|gt|gte]: Date }
  },
  aggregation_keys: [
    'domainId',
    'counterName',
    'counterType',
    'source',
    'timestamp'
  ]
});
```

That we internally map to an aggregation call and return flattened.

So... for a counter with this structure:
```JSON
{
  "domainId": "dashboard",
  "counterName": "<dashboard-id>",
  "counterType": "views",
  "source": "server"
}
```

1. If I only care about the grand total of my N-day retention period for all my dashboards, I'd call the API like
```TS
usageCounters.search({ 
  filters: {
    domainId: "dashboard",
    counterType: "views",
    source: "server",
  },
  aggregation_keys: []
});
```
And I'd get a response like
```TS
{
  counters: [
    { 
      domainId: "dashboard",
      counterType: "views",
      source: "server",
      count: 9999999,
    }
  ]
}
```
2. If I want the grand total for each dashboard, I'd call the API like
```TS
usageCounters.search({ 
  filters: {
    domainId: "dashboard",
    counterType: "views",
    source: "server",
  },
  aggregation_keys: [
    'counterName'
  ]
});
```
And I'd get a response like
```TS
{
  counters: [
    { 
      domainId: "dashboard",
      counterType: "views",
      source: "server",
      counterName: "dashboard-1"
      count: 10,
    },
    { 
      domainId: "dashboard",
      counterType: "views",
      source: "server",
      counterName: "dashboard-2"
      count: 9999989,
    }
  ]
}
```
3. If I want the histogram for a specific dashboard, I'd call the API like
```TS
usageCounters.search({ 
  filters: {
    domainId: "dashboard",
    counterType: "views",
    counterName: "dashboard-2",
    source: "server",
  },
  aggregation_keys: [
    'timestamp'
  ]
});
```
And I'd get a response like
```TS
{
  counters: [
    { 
      domainId: "dashboard",
      counterType: "views",
      counterName: "dashboard-2"
      source: "server",
      timestamp: "2024-07-01T00:00:00.000Z"
      count: 10,
    },
    { 
      domainId: "dashboard",
      counterType: "views",
      counterName: "dashboard-2"
      source: "server",
      timestamp: "2024-07-03T00:00:00.000Z"
      count: 9999989,
    }
  ]
}
```

The benefits of the aggregations are: 
* No pagination required
* Can be requested at any level

WDYT?

Reply:
It's an interesting proposal!

In the current use case, Anton still has to retrieve individual days counters to show them in the UI, so I went for the simplest approach possible, and let him aggregate counts on his side (saving 1 call in the process).

But I agree that it seems desirable to perform the aggregations on our side long term, makes for a more elegant API. Regarding pagination, when retrieving individual results you might still have plenty, but we currently circumvent this by allowing the `from: string` parameter. This way we can filter and only get counters that are more recent than a certain date (e.g. `now - 90d`).

Reply:
> Regarding pagination, when retrieving individual results you might still have plenty, but we currently circumvent this by allowing the `from: string` parameter. This way we can filter and only get counters that are more recent than a certain date (e.g. `now - 90d`).

AFAIK, the recommended way to paginate is via PIT for various reasons:

1. The `from` + `size` technique is limited to 10_000 entries: https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html (no matter if they are 10 pages of 1000 or 1000 pages of 10)
2. If updates occur in the process, the list will be resorted, so getting the 2nd+ pages may return previously fetched documents if new documents are indexed during the pagination.

AFAIK, we'll always retrieve all values queried, since the intention is not to show these values in a table that the user can paginate. So I don't think we're saving ourselves from any potential issues.

> Anton still has to retrieve individual days counters to show them in the UI

@Dosant, just FMI, will you retrieve all days for the histogram, and add them up to get the total? Or will the histogram be of the last 7 days and the total will account for the entire retention period (30d? 90d?)

> Let's discuss this offline!

Sure! Happy to meet when you're back :)

Reply:
@afharo, I retrieve last 90 days using `from`. Calculate "total" to display "Views in last 90 days" and bucket into weeks to display a weekly histogram

Reply:
After discussion with @afharo, we agreed that:
* We can leave aggregations for a later phase. I prepared the ground by encapsulating current filters under `filters` property.
* We must implement `PIT search`.

These changes have been implemented in https://github.com/elastic/kibana/pull/187665/commits/d8e86e48ee56bf21265f1cdfe9c9aa4c331c39c2